x/twap/listeners_test.go

```diff
@@ -3,35 +3,93 @@ package twap_test
 import (
 	"time"
 
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
 	"github.com/osmosis-labs/osmosis/v11/x/twap/types"
 )
 
-// TestCreatePoolFlow tests that upon a pool being created,
-// we have made the correct store entries.
-func (s *TestSuite) TestCreateTwoAssetPoolFlow() {
-	poolId := s.PrepareUni2PoolWithAssets(defaultUniV2Coins[0], defaultUniV2Coins[1])
-
-	expectedTwap, err := types.NewTwapRecord(s.App.GAMMKeeper, s.Ctx, poolId, denom0, denom1)
-	s.Require().NoError(err)
-
-	twap, err := s.twapkeeper.GetMostRecentRecordStoreRepresentation(s.Ctx, poolId, denom0, denom1)
-	s.Require().NoError(err)
-	s.Require().Equal(expectedTwap, twap)
-
-	twap, err = s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), denom0, denom1)
-	s.Require().NoError(err)
-	s.Require().Equal(expectedTwap, twap)
-}
-
-// TODO: Write test for create pool, swap, and EndBlock in same block, we use post-swap spot price in record.
-
-// Tests that after a swap, we are triggering internal tracking logic for a pool.
-func (s *TestSuite) TestSwapTriggeringTrackPoolId() {
-	poolId := s.PrepareUni2PoolWithAssets(defaultUniV2Coins[0], defaultUniV2Coins[1])
-	s.BeginNewBlock(false)
-	s.RunBasicSwap(poolId)
-
-	s.Require().Equal([]uint64{poolId}, s.twapkeeper.GetChangedPools(s.Ctx))
+// TestAfterPoolCreatedHook tests if internal tracking logic has been triggered correctly,
+// and the corerct state entries have been created upon pool creation.
+// This test includes test cases for swapping on the same block with pool creation.
+func (s *TestSuite) TestAfterPoolCreatedHook() {
+	two_asset_coin := defaultUniV2Coins
+	three_asset_coins := defaultThreeAssetCoins
+
+	tests := []struct {
+		name      string
+		poolCoins sdk.Coins
+		// if this field is set true, we swap in the same block with pool creation
+		runSwap bool
```

Comment:
Does it make sense to have `expectedTwapRecordCouns` or `expectedTwapRecords` here so that we can have more specific assertions? 

Currently, I think that most assertions we make depend on another method in the internal logic, and if both break, we might get a false positive

Reply:
Good point. In the latest commit I made it get the correct expected record as `types.NewTwapRecord(s.App.GAMMKeeper, s.Ctx, poolId, denomPairs0[i], denomPairs1[i])`, which we then test equality against. (As that is our definition of correct here)

Only thing I think that would be slipping is if:
* `types.NewTwapRecord` returned something wrong for an integration test failure,
* 0 records were created, and our expected logic missed this due to a bug in
  *  `types.GetAllUniqueDenomPairs(denoms)`failing. 
  *  denoms, err := s.App.GAMMKeeper.GetPoolDenoms(s.Ctx, poolId)

Reply:
I think trusting `types.NewTwapRecord` is fine, since thats our definition of correct for a new record. 

I also feel like `types.GetAllUniqueDenomPairs(denoms)` should be well tested / have no edge cases, but we could if we wanted add an 'expected num records' just to sanity check